Set curl CURLOPT_TIMEOUT to prevent processes being stuck after connecting to server

FelixJacobi opened the pull request:
E.g. connection was successful but server does not respond properly.

We encountered the situation that the HTTP connection was initially established but the cron script using the library hangs afterwards for an infinity time (something was wrong on TCP level). Debugging the issue showed something was wrong with the network link between both machines (fault on side of the ISP). This change adds a reliable timeout for the whole request process in cURL to prevent the library from hanging forever in future if such situation occurs.

I submitted this patch already against original upstream in https://github.com/bigbluebutton/bigbluebutton-api-php/pull/120. But it seems that this fork has more activity and the PR more likely will be accepted in near future.

Thank you for your time.

sualko commented:
Thanks for your pr. I think it's a great enhancement, but maybe 20 seconds are too short. How about 30 seconds?

SamuelWei commented:
> Thanks for your pr. I think it's a great enhancement, but maybe 20 seconds are too short. How about 30 seconds?

Maybe we could have the time as a parameter the users can adjust?

sualko commented:
> Maybe we could have the time as a parameter the users can adjust?

We could add it as parameters to the BigBlueButton, or use some static variables, so we don't have to pass it down. I'm wondering if there is no standard way of configuring curl via some php parameters.

joisarjignesh commented:
> > Maybe we could have the time as a parameter the users can adjust?
> 
> We could add it as parameters to the BigBlueButton, or use some static variables, so we don't have to pass it down. I'm wondering if there is no standard way of configuring curl via some PHP parameters.
yes I think we need to set up the third parameter as configuration on an array (optional) and also we need to set the default curl configuration array value

FelixJacobi commented:
> Thanks for your pr. I think it's a great enhancement, but maybe 20 seconds are too short. How about 30 seconds?

Would be fine for me, I guess. The initial approach was more like a qnd fix to repair our infrastructure ;-) .

> Maybe we could have the time as a parameter the users can adjust?

> We could add it as parameters to the BigBlueButton, or use some static variables, so we don't have to pass it down. I'm wondering if there is no standard way of configuring curl via some php parameters.

> yes I think we need to set up the third parameter as configuration on an array (optional) and also we need to set the default curl configuration array value

Yeah, sounds reasonable. But I am not really sure if this should belong into the `BigBlueButton` class directly. I would possibly pursue the idea of moving the direct use of cURL out of this class into a separate class for a transport (maybe with a suitable interface).

The values can then be injected via the constructor of this class instead of `BigBlueButton` and a suitable configured instance can be passed if necessary. It would further allow to replace the cURL implementation e.g. with a PSR HTTP client or similar if one e.g. wants to perform integration tests with the code.

If I am correct, it would be possible even without introducing any BC breaks.

sualko commented:
Maybe we can combine all ideas:

- increase the timeout to a usable value
- allow to inject transport class
- add params to curl transport

We just need someone who is willing to implement this :smile: Volunteers?

FelixJacobi commented:
> We just need someone who is willing to implement this smile Volunteers?

I can work on it this weekend.

FelixJacobi commented:
Hi everyone,

A first draft of HTTP transports is now included in this PR. I moved the original cURL integration into a `CurlTransport`. I also added a `TransportInterface` and transports for Symfony HTTP Client and PSR18 HTTP Client as quite common and spread libraries (possibly could be moved into separate Composer packages later).

Furthermore, I am planning further refinements (e.g. add tests) during the next days. The code was also not yet tested in an environment with a real BBB server. Comments already welcome.

FelixJacobi commented:
I also addresses the suggestion by @sualko and raised the default request timeout in `CurlTransport`  to 30 seconds.

FelixJacobi commented on src/Http/Transport/Bridge/PsrHttpClient/PsrHttpClientTransport.php:
```diff
@@ -0,0 +1,116 @@
+    {
+        if ('' !== $payload = $request->getPayload()) {
+            $psrRequest = $this->requestFactory->createRequest('POST', $request->getUrl());
+            $psrRequest = $psrRequest->withBody(Psr7Stream::createFromString($payload));
```
TODO: Replace class copied from Guzzle with `StreamFactory` from PSR-18.

sualko commented on src/Http/Transport/Bridge/PsrHttpClient/PsrHttpClientTransport.php:
```diff
@@ -0,0 +1,134 @@
+declare(strict_types=1);
+
+/**
+ * BigBlueButton open source conferencing system - https://www.bigbluebutton.org/.
```
Since the code is not from BBB, I think we can use the recommended lgpl header.

FelixJacobi replied:
I hope I used the recommended header you meant. On gnu.org, I only found a definitive header for the GPL, not LGPL.

sualko commented on LICENSE.3rdparty:
```diff
@@ -0,0 +1,26 @@
+Portions of this code were took from the Guzzle PHP HTTP client project (https://guzzlephp.org/en/stable/) and are available under the following terms:
```
It would be great if this license notice could got to the related files, because currently it's not clear which parts are copied. I'm also not sure if this are "substantial portions of the Software".

sualko replied:
I think the correct url is https://docs.guzzlephp.org/en/stable/

FelixJacobi replied:
> I'm also not sure if this are "substantial portions of the Software".

I am not, too. As I am not lawyer, I would not try to interpret these conditions and always add the hint to used code ;-) .

> It would be great if this license notice could got to the related files, because currently it's not clear which parts are copied.

Done.

sualko reviewed (changes requested):
I did not test those changes, but they look awesome and should be backward compatible.

We should just discuss the license stuff.

FelixJacobi commented:
From my side, this PR should be finished now. I incorporated some further minor improvements and included tests (even for the `CurlTransport` with a very special solution).

sualko commented:
> From my side, this PR should be finished now. I incorporated some further minor improvements and included documentation and tests (even for the CurlTransport with a tricky but good working solution).

Great. Do you have time to merge the master branch, before we do a review?

FelixJacobi commented:
> > From my side, this PR should be finished now. I incorporated some further minor improvements and included documentation and tests (even for the CurlTransport with a tricky but good working solution).
> 
> Great. Do you have time to merge the master branch, before we do a review?

Sure. Will do it this evening or tomorrow.

FelixJacobi commented:
> > From my side, this PR should be finished now. I incorporated some further minor improvements and included documentation and tests (even for the CurlTransport with a tricky but good working solution).
> 
> Great. Do you have time to merge the master branch, before we do a review?

@sualko Done.

FelixJacobi commented:
I noticed that the PHP 7.2 pipeline seems to permanently fail. The failure seems to be related to Coveralls. Anyone knows what to do?

SamuelWei commented:
> I noticed that the PHP 7.2 pipeline seems to permanently fail. The failure seems to be related to Coveralls. Anyone knows what to do?

https://github.com/littleredbutton/bigbluebutton-api-php/pull/60/commits/e86e8a211eaba66acdc3d799c4011503e565f3c2

FelixJacobi commented:
> > I noticed that the PHP 7.2 pipeline seems to permanently fail. The failure seems to be related to Coveralls. Anyone knows what to do?
> 
> [e86e8a2](https://github.com/littleredbutton/bigbluebutton-api-php/commit/e86e8a211eaba66acdc3d799c4011503e565f3c2)

Hm, seems already to be applied in this branch.

FelixJacobi commented:
Regarding #64: Would it help if someone with write permissions to this repository made a copy of the branch and create a new PR?

sualko commented:
Should be no big issue to merge this pr from command line, but until now I had no time to look at it. Sorry.

sualko reviewed (commented):
Looks really nice. Enough tests and everything is documented nicely. Well done.

I just need to check it out and run the integration tests. Maybe @SamuelWei has also time to review this pr.

sualko commented:
@SamuelWei can you give a time frame until you have reviewed this? I think after this is merged we can release v3.

SamuelWei commented:
> @SamuelWei can you give a time frame until you have reviewed this? I think after this is merged we can release v3.

I agree, that would be a big improvement and change that we should release as v3.4.0

Giving a timeframe isn't easy, as my daily schedule gets changed quite frequently. I hope I'll be able to have a first deeper look into the PR tomorrow and next week. Today I just scrolled through all the changes and liked the idea and implementation.

SamuelWei commented on src/Http/Transport/CurlTransport.php:
```diff
@@ -0,0 +1,196 @@
+        // @codeCoverageIgnoreEnd
+        $cookieFilePath = stream_get_meta_data($cookieFile)['uri'];
+
+        $options = ArrayHelper::mergeRecursive(
```
If you set the http header for Content-type or Content-length as shown in this example, you get duplicates

```php
$request   = new TransportRequest('http://localhost:8057/dump.php', 'FOO', 'application/xml');
$transport = new CurlTransport([CURLOPT_HTTPHEADER => ['X-Foo: Bar', 'X-Bar: Foo','Content-length: 5']]);
$response = $transport->request($request);
```


```php
array(5) {
    [0] =>
    string(29) "Content-type: application/xml"
    [1] =>
    string(17) "Content-length: 3"
    [2] =>
    string(10) "X-Foo: Bar"
    [3] =>
    string(10) "X-Bar: Foo"
    [4] =>
    string(17) "Content-length: 5"
  }
```

SamuelWei replied:
A solution could be to parse all headers to an array, merge/overwrite and reassemble.

FelixJacobi replied:
Good catch! I guess it would be fine to implement that.

FelixJacobi replied:
I added a specialized header util  for that case plus a test of the scenario you have described.

FelixJacobi commented:
Should be ready now. I added integration tests for the new transports like there already exists for the original cURL one.

sualko commented:
I think we can merge this now. If we find bugs we can fix them afterwards. Before I release v3, I want to test this in my application.

@FelixJacobi thanks for this awesome job.

sualko commented:
Correction: We have to release v4, since we are already on 3.x :see_no_evil:.

@FelixJacobi if you like you can create another pr to add yourself to the list of authors in the composer file.

FelixJacobi commented:
> @FelixJacobi if you like you can create another pr to add yourself to the list of authors in the composer file.

Sure, thanks. Will do this in #67.
